[OGUI-1722] Runs mode frontend

mariscalromeroalejandro opened the pull request:
#### I have JIRA issue created
- [x] branch and/or PR name(s) includes JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected
- [x] FLP integration tests were ran successful

In this PR, a new option to enter 'run mode' is displayed when a user inputs a run number. For the moment, it is only allowed on the 'object tree' page:

- It is disabled when no run number is entered.
- When selected, the application enters 'run mode', and the number and status are displayed.
- Paths are retrieved periodically with a manually set timeout (this needs to be changed) if the run is ongoing.
- The status of the run can be: 'ongoing', 'ended', 'not found', or 'unknown'.

The goal is to extend this feature to the rest of the pages, not just the object tree.

INTERVALS MANAGEMENT IS PROPERLY MANAGED IN PR https://github.com/AliceO2Group/WebUi/pull/2982
In this PR, we handle it by resetting the run mode status because it is only applied to the object tree page

Commit:
feat: add run retrieval functionallity

Commit:
[OGUI-1690] compressed

Commit:
style: add css classes

Commit:
feat: perstist filters through page reloads and page changes wip

Commit:
test: add deticated filter integration tests

Commit:
test: adjust tests to work with new ui changes

Commit:
refactor: unify objectTree and sideTree

Commit:
feat: add filterToggle element

Commit:
feat: put filterToggleButton in page headers

Commit:
feat: add filterElement to header and remove existing filters

Commit:
feat: add filterClear button to filterElement

Commit:
feat: remove filter dependency from models and move it to services

Commit:
feat: optionally disable objectTree filters from filters

Commit:
chore: remove filterToFromUrlParams.js

Commit:
style: add Id's for tests

Commit:
feat: create runStatus.enum

Commit:
feat: add runStatusFetch functionallity for filterService

Commit:
refactor: improve error message

Commit:
feat: implement interval for getObjectById

Commit:
feat(objectService): add runcache

Commit:
feat: add status polling functionallity for getObjects

Commit:
feat: implement interval for getObjectContent

Commit:
test: create tests for BookeepingService and ObjectController

Commit:
chore: fix documentation

Commit:
chore: decapitalise RunNumber for function parameters

Commit:
[OGUI-1689] compressed

Commit:
Merge branch 'dev' of github.com:AliceO2Group/WebUi into feature/QCG/OGUI-1690/automatically-fill-in-run-number-on-applicable-pages-v2

Commit:
fix: notify on clearing input filters.

Commit:
feat: create BaseViewModel abstract class

Commit:
refactor(FilterModel): replace visble with isVisible

Commit:
refactor(Layout): change applyTreeFilter to shoutldApplyTreeFilter

Commit:
fix: make all page models extend BaseViewModel

Commit:
feat: reload objects when clearing the page.

Commit:
fix: don't force objectView to fetch by name

Commit:
fix: allow objectView to swap between versions while maining reference to commits

Commit:
fix: do not update url when setting filters with dropdown.

Commit:
chore: move applyFilters button down

Commit:
Merge branch 'feature/QCG/OGUI-1690/automatically-fill-in-run-number-on-applicable-pages-v2' of github.com:AliceO2Group/WebUi into feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs

NarrowsProjects committed:
Merge branch 'dev' into feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs

Commit:
feat: define runStatusRefreshRate in config-default

Commit:
chore(ObjectController): update the jsdoc for updateAndCheckStatus

Commit:
fix: fix eslint issue

Commit:
refactor: add warning log in retrieveRunStatus

Commit:
feat: return Invalid if there is no connection to bookkeeping.

Commit:
chore: put documentation back above getObjectById (Again)

Commit:
chore: update documentation with new bookkeeping fields

Commit:
feat: move intervals from bookkeeping to filterService

Commit:
rename bookkeepingServiceTest

Commit:
chore: updte example token

Commit:
undo test comments

Commit:
chore: remove unused triggerfilters from models

Commit:
chore(ObjectController): improve error message

NarrowsProjects committed:
Merge branch 'dev' into feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs

mariscalromeroalejandro committed:
move run monitoring to new service

mariscalromeroalejandro committed:
add tests, update controller tests

mariscalromeroalejandro committed:
check status only if needed. improve jsdoc

mariscalromeroalejandro committed:
undo skipping drontend tests

graduta committed:
Rename RunStatus in line with BKP namings

graduta committed:
Update RunMode to check for runMode only

graduta committed:
Update Object Controller

graduta committed:
More updates to revert refactor

graduta committed:
Add interval only if config bkp is defined

mariscalromeroalejandro committed:
adapt object controller test

mariscalromeroalejandro committed:
bkp tests: invalid status -> not found (run status)

mariscalromeroalejandro committed:
new QcObjectService test suite

mariscalromeroalejandro committed:
run mode service test suite

mariscalromeroalejandro committed:
fix filter service test suite

mariscalromeroalejandro committed:
fix: initialize intervals to include runModeService

mariscalromeroalejandro committed:
fix: error destructuring run number if not provided

mariscalromeroalejandro committed:
runs mode checkbox component

mariscalromeroalejandro committed:
adapt filters

mariscalromeroalejandro committed:
adapt header to include runs mode component

mariscalromeroalejandro committed:
qcobject model and service update

mariscalromeroalejandro committed:
object tree header update. model update

mariscalromeroalejandro committed:
move parseObjects as a utility

mariscalromeroalejandro committed:
include run status in response

mariscalromeroalejandro committed:
accept 'inRunMode' as a query parameter

mariscalromeroalejandro committed:
fix tests

mariscalromeroalejandro committed:
Merge branch 'feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs' into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
add coverage for parseObjects method

mariscalromeroalejandro committed:
Merge branch 'feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs' into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
refactor styles

mariscalromeroalejandro committed:
enter/exit run mode refactoring

mariscalromeroalejandro committed:
revert css format changes

mariscalromeroalejandro committed:
hide filters in run mode only in object tree page

mariscalromeroalejandro committed:
fix object tree tests

mariscalromeroalejandro committed:
fix failing test

mariscalromeroalejandro committed:
test to check availability of runs mode checkbox

mariscalromeroalejandro committed:
avoid error when runNumber is 0 by checking undefined explicitly

mariscalromeroalejandro committed:
Merge branch 'feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs' into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
more tests for runs mode (using id property)

mariscalromeroalejandro committed:
better checking for run number in controller

mariscalromeroalejandro committed:
check if run number is an integer to activate runs mode

mariscalromeroalejandro committed:
fix object tree tests

mariscalromeroalejandro committed:
Merge branch 'dev' into feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs

mariscalromeroalejandro committed:
fix type validation and code structure

mariscalromeroalejandro committed:
test: improve BkpService test coverage

mariscalromeroalejandro committed:
refactor: improve FilterService constructor and async initialization

mariscalromeroalejandro committed:
refactor cache refresh method

mariscalromeroalejandro committed:
fix lint errors

mariscalromeroalejandro committed:
update version ca-certificates

mariscalromeroalejandro committed:
fix double semicolom

mariscalromeroalejandro committed:
FilterModel.js - Missing this. references

mariscalromeroalejandro committed:
global model access / add jsdoc

mariscalromeroalejandro committed:
allow testing in Bookkeping Service

mariscalromeroalejandro committed:
use camelCase bookkeepingService

mariscalromeroalejandro committed:
protect against not found object. fix typo

mariscalromeroalejandro committed:
add test to check run number format if in run mode

mariscalromeroalejandro committed:
fix eslint: trailing spaces not allowed

mariscalromeroalejandro committed:
Merge branch 'feature/QCG/OGUI-1721/Fetch-run-status-and-object-data-on-interval-for-active-runs' into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
Merge branch 'dev' into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
fix eslint

mariscalromeroalejandro committed:
split runs mode tests

mariscalromeroalejandro committed:
Merge branch 'dev' into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
Fixed object tree header: Changed model.inRunMode to qcObject.model.inRunMode

mariscalromeroalejandro committed:
test: changed the checkbox selection to use input type instead of id

graduta commented on QualityControl/public/Model.js:
```diff
@@ -78,6 +78,11 @@ export default class Model extends Observable {
     this.router.observe(this.handleLocationChange.bind(this));
 
+    //Run mode
+    this._inRunMode = false; // Whether the application is in run mode or not
```
Would this not make a better addition to the FilterModel as it is tied to the user adding a runNumber in the input?

graduta commented on QualityControl/public/Model.js:
```diff
@@ -78,6 +78,11 @@ export default class Model extends Observable {
 
+    //Run mode
+    this._inRunMode = false; // Whether the application is in run mode or not
+    this._runNumber = null; // Run number to be used in run mode
```
This is already part of the filter to use, right?

graduta commented on QualityControl/public/Model.js:
```diff
@@ -241,7 +246,7 @@ export default class Model extends Observable {
         this.page = 'objectTree';
         setBrowserTabTitle('QCG-Tree');
-        this.object.loadList();
+        await this.object.loadList();
```
Is there a reason we would like to wait for this before proceeding with loading rest of the page?

graduta commented on QualityControl/public/Model.js:
```diff
@@ -362,4 +367,58 @@ export default class Model extends Observable {
+    return this._inRunMode;
+  }
+
+  enterRunMode() {
```
Missing documentation

graduta commented on QualityControl/public/Model.js:
```diff
@@ -362,4 +367,58 @@ export default class Model extends Observable {
+    this.notify();
+  }
+
+  exitRunMode() {
```
Missing documentation

graduta commented on QualityControl/public/Model.js:
```diff
@@ -362,4 +367,58 @@ export default class Model extends Observable {
+
+  /**
+   * Sets the current run status and triggers a notification update.
+   * @param {string} value - The new run status (e.g., 'ONGOING', 'COMPLETED').
```
Where is `COMPLETED` coming from? As far as I can see from your previous PR status is one of
```
  ENDED: 'ENDED',
  ONGOING: 'ONGOING',
  NOT_FOUND: 'NOT_FOUND',
```

graduta commented on QualityControl/public/common/filters/model/FilterModel.js:
```diff
@@ -141,4 +142,60 @@ export default class FilterModel extends Observable {
+        if (this.model.runStatus !== 'ONGOING') {
+          this._clearRunsModeInterval();
+        }
+        // TODO: Should be provided in config file ??
```
Yes, please make a ticket for this

graduta commented on QualityControl/public/common/filters/model/FilterModel.js:
```diff
@@ -141,4 +142,60 @@ export default class FilterModel extends Observable {
+   */
+  async _manageRunsModeInterval(baseViewModel) {
+    this._clearRunsModeInterval();
+    if (this.model.runStatus === 'ONGOING') {
```
It would be better to use the RunStatus enum rather than hard-coded strings

graduta commented on QualityControl/public/common/filters/model/FilterModel.js:
```diff
@@ -141,4 +142,60 @@ export default class FilterModel extends Observable {
+      this._runsModeInterval = setInterval(async () => {
+        await baseViewModel.triggerFilter(true);
+        this.notify();
+        if (this.model.runStatus !== 'ONGOING') {
```
Same as above

graduta reviewed (changes requested):
First prototype looking good. I messaged you a bit more feedback on user friendliness. Keep up the good work!

mariscalromeroalejandro committed:
delegates run mode operations to FilterModel

mariscalromeroalejandro committed:
run status enum

mariscalromeroalejandro committed:
center object tree title

mariscalromeroalejandro committed:
exit runMode if location changes

mariscalromeroalejandro committed:
new run mode header (remove checkbox)

mariscalromeroalejandro committed:
remove run mmode checkbox tests

mariscalromeroalejandro committed:
adapt tests

mariscalromeroalejandro committed:
fix warning: Superfluous trailing arguments

mariscalromeroalejandro committed:
Merge branch 'dev' into feature/QCG/OGUI-1722/runs-mode-frontend

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+      if (runNumber === 'N/A') {
+        return;
+      }
+      await navigator.clipboard.writeText(runNumber);
```
`navigator.clipboard`, as per documentation (https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard) is only available in secure contexts.
That means that this particular line will throw an error. Please have a look in the codebase on how we implement the `copy` functionality

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+  function renderTitle() {
+    return h('span.run-mode-title', {
+      id: 'run-mode-title',
+      title: 'Run Mode filters objects by a specific run number and monitors run status.' +
```
I think this should really be moved to the `icon` pop-up

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+    id: 'run-mode-header',
+  }, [
+    renderRunModeInfo(model, filterModel),
+    h('.flex-grow'),
```
What is the purpose of this empty div?

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+      id: 'run-mode-info',
+    }, [
+      renderTitle(),
+      renderRunNumber(model.runNumber),
```
RunNumber is already provided by the FilterModel. Could we not use it from there and remove this getter from the Model.js?

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+    }, [
+      renderTitle(),
+      renderRunNumber(model.runNumber),
+      renderRunStatus(model, filterModel),
```
Is there a need for sending an entire `model` only for `runStatus`?

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+   */
+  function renderRunNumber(runNumber = 'N/A') {
+    const copyToClipboard = async () => {
+      if (runNumber === 'N/A') {
```
Do we need a hardcoded string `N/A` across the codebase?
The filter has the runNumber already as a value or `undefined`

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+        id: 'run-number-value',
+        title: runNumber === 'N/A' ? 'No run number available' : 'Click to copy run number',
+        onclick: copyToClipboard,
+        style: runNumber !== 'N/A' ? 'cursor: pointer; user-select: none;' : 'cursor: default;',
```
Please use classes for styling. These `h()` elements allow you to pass classes separated by a comma as a string. There are lots of examples in the codebase.

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+    return h('div', [
+      h('div', ...Object.values(RunStatus).map((status) =>
+        h('label', {
+          style: 'transition: transform 0.2s;',
```
This transformation implies an action that can be taken by the user.
What is the action supposed to be in this `legend`?

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+   */
+  function getStatusClass(status) {
+    if (!status) {
+      return 'text-muted';
```
Why not using an existing color or class such as `gray`?

graduta commented on QualityControl/public/common/runModeHeader.js:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    switch (status.toLowerCase()) {
+      case 'ended':
```
You have the runStatus enum introduced above in the PR. Perhaps you should be using that rather than hardcoded strings?
Moreover, we have status CSS classes declared in the code base already (`bootstrap.css`) such as `success`, `primary`, etc.

graduta reviewed (changes requested):
I am not a fan of all the duplication of enter/exit/isactiverunmode etc functions.
These do not belong to the Model.js but in the interest of advancing we will continue with this just for this PR.
In the next task, when adding the run mode to rest of the pages, please consider improving this with its own model and integration in FilterModel as it is tied to it

mariscalromeroalejandro committed:
move all runs mode logic to filter model

mariscalromeroalejandro committed:
clear filtersmodel

mariscalromeroalejandro committed:
refactor styles

mariscalromeroalejandro committed:
remove global model from runModeHeader

mariscalromeroalejandro committed:
wait for filter service to initialize

mariscalromeroalejandro committed:
[OGUI-1741] Remove filters layout list about page (#2978)

This PR removes the filter components from the About page and Layout List page as they are not relevant for these pages and were causing confusion in the user experience.

- About page: Removed filter panel from `aboutViewHeader.js` (filters don't apply to static information content)
- Layout List page: Removed filter panel from `LayoutListHeader.js` (layout search functionality is sufficient for this page)
- Updated filter persistence test to only validate filter state between relevant pages

mariscalromeroalejandro committed:
new api endpoint to retrieve run status

mariscalromeroalejandro committed:
use new endpoint to retrieve run status

mariscalromeroalejandro committed:
move runs mode component below the header

mariscalromeroalejandro committed:
Merge branch 'feature/QCG/OGUI-1722/runs-mode-frontend' of https://github.com/AliceO2Group/WebUi into feature/QCG/OGUI-1722/runs-mode-frontend

mariscalromeroalejandro committed:
reset runs mode except objectTree

mariscalromeroalejandro committed:
remove parameters of filter model passed to the layouList and about page view

mariscalromeroalejandro committed:
add jsdoc

mariscalromeroalejandro committed:
use runStatus from common/library

mariscalromeroalejandro committed:
fix error handling in run status endpoint

mariscalromeroalejandro committed:
make validation in filter controller

mariscalromeroalejandro committed:
fix filter persistance

mariscalromeroalejandro committed:
filter controller/service tests

mariscalromeroalejandro committed:
api get run status tests

mariscalromeroalejandro committed:
remove run status from api response

mariscalromeroalejandro committed:
fix jsdoc warnings

mariscalromeroalejandro committed:
fix import case sensitivity and optimize run status fetching

mariscalromeroalejandro committed:
save tree state after refreshing in run mode when a run is ongoing

mariscalromeroalejandro committed:
run number dto validation

mariscalromeroalejandro committed:
tests for runs mode in object tree view
